### PR TITLE
DRY refactoring

### PR DESCRIPTION
**Pull request by charles-cowart**
Goal is to refactor code so that there is minimal duplication, easily extensible to enable metatranscriptomics and future Assay types, and breakout all of the functionality from the main plugin function so they can be tested properly.

### CODE REVIEW

File: .gitignore
Comment (antgonza) [inline .gitignore:132]: why delete this?

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:16]: Should this just be Amplicon? The AmpliconStep description might be too technical, what do you think?

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:20]: Is this hardcoded? Just asking but if is maybe adding a note about it would be good.

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:21]: This note makes sense but it would be better if it was clear that this is by design and historically (EMP primers)

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:25]: Something that concerns me about this is that this looks like it's completely different from other pipelines, is that right? I thought the idea was to merge it. Note that I haven't read any code below.

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:81]: Should this be part of self/super() and not have this here?

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:86]: Same as comment above, this feels like the other pipeline is going to have the same methods

File: qp_klp/AmpliconStep.py
Comment (antgonza) [inline qp_klp/AmpliconStep.py:98]: Same here, this should be part of the main class, no?

File: qp_klp/MetagenomicStep.py
Comment (antgonza) [inline qp_klp/MetagenomicStep.py:28]: Yup, this looks like duplicated code between the two classes.

Comment (antgonza) [reply qp_klp/MetagenomicStep.py:28]: Going to skip to the next section.

File: qp_klp/klp.py
Comment (antgonza) [inline qp_klp/klp.py:184]: Looking at this and the code in Step is clear that this is not a step but a controller, similar to what you used to have of Pipeline; is that right? If yes, why the rename?

File: qp_klp/tests/test_amplicon_step.py
Comment (antgonza) [inline qp_klp/tests/test_amplicon_step.py:20]: why is this part of the object if it's only use within the setUp? Is the plan to reuse it somewhere else?

File: qp_klp/tests/test_amplicon_step.py
Comment (antgonza) [inline qp_klp/tests/test_amplicon_step.py:39]: metagenomic_pipeline? ?

File: qp_klp/tests/test_amplicon_step.py
Comment (antgonza) [inline qp_klp/tests/test_amplicon_step.py:54]: `AmpliconStep(metagenomic_pipeline`? It can't be both, right?

**Review by antgonza — CHANGES_REQUESTED** [review]
Thank you @charles-cowart, I have some concerns about the possibility of duplication of code; please let me know what you think.

File: .gitignore
Comment (charles-cowart) [reply .gitignore:132]: For this refactor I added the configuration.json file that is also a part of the mg-scripts repo into this repo. This necessitated removing this line from .gitignore so it could be added. This information is needed for running tests. In the past we were able to avoid adding configuration.json by hardcoding the data-structure in the tests. I wanted to avoid this in the new tests.

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py:16]: Your call. It's more wordy but I think it was convention for so long to name subclasses like this that I got used to it. Amplicon would be nice and shorter too; not sure if it will get confusing with other usages of the word 'amplicon' in the code. What do you propose?

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py:20]: No less than previously in qp-klp but yeah, the knowledge to look for 'bcl2fastq' for Amplicon and 'bcl-convert' for Metagenomics is hardcoded. I can think about how to do that in a more thoughtful way, or listen to suggestions. For now I'll add a comment.

File: qp_klp/tests/test_amplicon_step.py
Comment (charles-cowart) [reply qp_klp/tests/test_amplicon_step.py:54]: I am indeed cheating a little. I'm testing that missing parameters raise the right Errors and I know that due to the ordering of checks, it doesn't matter whether I use a metagenomic pipeline or an amplicon pipeline. I'll use an amplicon pipeline to avoid confusion.

**Comment by charles-cowart** [comment]
@antgonza I've pushed a commit of quick updates based on your feedback, thanks! I haven't responded to all of your concerns yet, but I will shortly.

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py]: The Jobs classes encapsulate all of the functionality we saw in the original shell scripts along with things we expected we'd want to have at that time. Since the actual creation and use of the Jobs classes was in qp-klp, it became the natural place to do work needed for each step and stick in additional features. A good example of this is generating failed sample records. We added it to the metagenomic pipeline but we removed it for the amplicon pipeline since we don't demultiplex the sample-names until the data is passed to Qiita.

Hence, each version of the pipeline has something they need to do at each stage; many of the stages share many of the same steps, but some need to do things before or after. In a few cases such as the QC stage of the amplicon pipeline, we need to skip adapter trimming and human filtering entirely and instead move and rename raw fastq files so they're in the form and location expected by the FastQC step.

AmpliconStep, MetagenomicStep, and future Step classes each encapsulate all the work that needs to be done at each stage. The elements common to each version are moved into the base class and called by the child classes instead of having the same code duplicated in each version.

In some cases where there's no difference between the two pipelines, they'll both just call the base class version of the step. In many cases, there'll be work that needs to be done or after the base call like generating failed sample reports in the metagenomics pipeline or setting up the ConvertJob object to use bcl2fastq instead of bcl-convert and creating a dummy sample-sheet for amplicon.

That's perhaps why the files look similar, even though there isn't much duplication of actual lines of code. The reason why the base class versions of the functions begin with '_' e.g. _quality_control() is because often the child versions have a different set of parameters and Python requires overriding methods to have the same parameter list. I could get around that with kwarg but I personally like the clarity of position parameters.

File: .gitignore
Comment (antgonza) [reply .gitignore:132]: Does this mean that now we have 2 config files one provided by mg-scripts and another by this repo?

File: qp_klp/AmpliconStep.py
Comment (antgonza) [reply qp_klp/AmpliconStep.py]: Just to be clear, I'm talking about the ValueError message, my concern is that a user will not know the difference between AmpliconStep/Amplicon and might confuse them, what do you think?

File: qp_klp/AmpliconStep.py
Comment (antgonza) [reply qp_klp/AmpliconStep.py]: Just to be clear, the point of this PR/work is to make this repo more python like (including nicer tests and remove duplication of code) vs. keeping the structure of the original shell script. 

Anyway, I understand your line of thought here: an object for each "Step" and have shared code in the base clase; which I guess is fine. Thank you for the explanation.

File: qp_klp/tests/test_amplicon_step.py
Comment (antgonza) [reply qp_klp/tests/test_amplicon_step.py]: cheating a little or a lot, sound wrong, right? Thank you for making the change.

**Review by antgonza — COMMENTED** [review]
Thank you for the comments before. Now, a general question, if you are "extracting" the general functions from Amplicon/Metagenomic Step and adding them to the base class Step, does it really make sense to have a klp_util.py file? I mean basically you have to access method X in Amplicon/Metagenomic Step, that will call X method in Step and then do X in klp_util.py; no? Would it be better to remove one step and just go either Amplicon/Metagenomic Step -> klp_util.py or Amplicon/Metagenomic Step -> Step and remove klp_util.py?

File: .gitignore
Comment (charles-cowart) [reply .gitignore:132]: Yes. The alternative is to keep the data-structure as a const in the test code.

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py]: I see where you're coming from but it could be argued that Error messages are as much for the developer as the user. The stack-traces we show in Qiita to the user have little value to the user, but convey that an error occurred and provide valuable insight to the developer.

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py]: Moved into the base _generate_reports() class. Thanks!

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py]: Not in this case, as we want to change the binary name for Amplicon runs from 'seqpro' to 'seqpro_mf'.

File: qp_klp/AmpliconStep.py
Comment (charles-cowart) [reply qp_klp/AmpliconStep.py]: Not sure if you mean the method above generate_commands() or generate_commands() itself, but in both cases each version needs different things done after the base call. In the case of generate_commands(), 'ConvertJob/Logs' needs to be included in the reports-ConvertJob.tgz tarball, whereas that directory is not created by bcl2fastq and therefore the Amplicon version.

File: qp_klp/MetagenomicStep.py
Comment (charles-cowart) [reply qp_klp/MetagenomicStep.py]: In the case of convert_bcl_to_fastq(), one uses configuration info for bcl2fastq while the other uses bcl-convert. Also the metagenomic version generates a failed sample report afterward. Not sure where you're seeing the duplication.

File: qp_klp/tests/test_amplicon_step.py
Comment (charles-cowart) [reply qp_klp/tests/test_amplicon_step.py:20]: Good catch! No reason, just an artifact of having been cut and pasted from older code.

File: qp_klp/klp.py
Comment (charles-cowart) [reply qp_klp/klp.py:184]: Not like Pipeline. 

> Thank you for the comments before. Now, a general question, if you are "extracting" the general functions from Amplicon/Metagenomic Step and adding them to the base class Step, does it really make sense to have a klp_util.py file? I mean basically you have to access method X in Amplicon/Metagenomic Step, that will call X method in Step and then do X in klp_util.py; no? Would it be better to remove one step and just go either Amplicon/Metagenomic Step -> klp_util.py or Amplicon/Metagenomic Step -> Step and remove klp_util.py?

We're definitely on the same page. The idea was to incorporate klp_util.py into the base Step class. I left them as is for the time being, just because so many other things had changed already.

**Comment by charles-cowart** [comment]
> Thank you for the comments before. Now, a general question, if you are "extracting" the general functions from Amplicon/Metagenomic Step and adding them to the base class Step, does it really make sense to have a klp_util.py file? I mean basically you have to access method X in Amplicon/Metagenomic Step, that will call X method in Step and then do X in klp_util.py; no? Would it be better to remove one step and just go either Amplicon/Metagenomic Step -> klp_util.py or Amplicon/Metagenomic Step -> Step and remove klp_util.py?

We're definitely on the same page. The idea was to incorporate klp_util.py into the base Step class. I left them as is for the time being, just because so many other things had changed already.

**Comment by antgonza** [comment]
> We're definitely on the same page. The idea was to incorporate klp_util.py into the base Step class. I left them as is for the time being, just because so many other things had changed already.

I think this will be the only blocker before merging this initial PR and this would be the base for future ones, could you do this now before waiting for a better time?

File: qp_klp/Amplicon.py
Comment (antgonza) [inline qp_klp/Amplicon.py:72]: Not a fan of this name, why fake?

File: qp_klp/Amplicon.py
Comment (antgonza) [inline qp_klp/Amplicon.py:78]: Why not join all these if statements and only loop the list once?

File: qp_klp/Step.py
Comment (antgonza) [inline qp_klp/Step.py:31]: is this really failed? At this stage it looks like all samples; it might be confusing for other devs.

Comment (antgonza) [reply qp_klp/Step.py:31]: One option is to have: self.samples (all samples) and then another one with self.samples_used and self.failed can be the difference between both.

File: qp_klp/Metagenomic.py
Comment (antgonza) [inline qp_klp/Metagenomic.py:16]: Should this be metatranscriptomic too? Note that this might be concerning as there are several places where this check is done, perhaps having a global set or something might be better; it could be part of the object too.

File: qp_klp/Metagenomic.py
Comment (antgonza) [inline qp_klp/Metagenomic.py:106]: This is another case about Metatranscriptomic

File: qp_klp/klp.py
Comment (antgonza) [inline qp_klp/klp.py:185]: metatransciptomic??

**Review by antgonza — COMMENTED** [review]
Some questions.

File: qp_klp/Amplicon.py
Comment (charles-cowart) [reply qp_klp/Amplicon.py:72]: It's faked in the sense that Amplicon doesn't invoke QCJob but the subsequent step expects a QCJob style layout. Rather than alter QCJob we simply copy the raw output from ConvertJob into something that looks like QCJob output.

Comment (charles-cowart) [reply qp_klp/Amplicon.py:72]: I'll come up with an alternative name.

Comment (antgonza) [reply qp_klp/Amplicon.py:72]: I think output_folder is enough, no?

Comment (charles-cowart) [reply qp_klp/Amplicon.py:72]: same. just changed it to that, actually.

File: qp_klp/Step.py
Comment (charles-cowart) [reply qp_klp/Step.py]: I changed the name from self.failed to self.sample_state - you're right, we're initializing the member with the entire list of samples but we don't know that any of them are failed yet.

File: qp_klp/tests/test_amplicon_step.py
Comment (antgonza) [inline qp_klp/tests/test_amplicon_step.py:25]: Do you really need a variable for this? Can the `{}` be passed in the function?

**Review by antgonza — COMMENTED** [review]
A few more comments.

**Review by antgonza — COMMENTED** [review]
A few more comments.

File: qp_klp/Amplicon.py
Comment (charles-cowart) [reply qp_klp/Amplicon.py:77]: I think it might be better to keep the separate, because you want x in isfile(x) in the second one to be the full path created by join(raw_fastq_files_path, x) in the first one and if we merged them that would mean we'd have something like:

job_output = [join(raw_fastq_files_path, x) for x in listdir(raw_fastq_files_path) if isfile(join(raw_fastq_files_path, x)) and x.endswith('fastq.gz') and not basename(x).startswith('Undetermined')]]

The join() has to be present twice.

File: qp_klp/Amplicon.py
Comment (charles-cowart) [reply qp_klp/Amplicon.py:90]: It's an easy enough change on this side of things, and a little bit of refactoring for metapool package. I can make the change by the end of the week in between other things.

File: qp_klp/Metagenomic.py
Comment (charles-cowart) [reply qp_klp/Metagenomic.py:36]: Sure, no problem!

**Pull request merged**
